eth/beacon/state_machines/base.py

```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
+        processing_active_state = cls.compute_per_block_transtion(
+            crystallized_state,
+            active_state,
+            block,
+            chaindb,
+            config.CYCLE_LENGTH,
+        )
+
+        # Process per cycle state changes (CrystallizedState and ActiveState)
+        processed_crystallized_state, processed_active_state = cls.compute_cycle_transitions(
+            crystallized_state,
+            processing_active_state,
+        )
+
+        # Return the copy
+        result_block = block.copy()
+        return result_block, processed_crystallized_state, processed_active_state
+
+    @classmethod
+    def compute_per_block_transtion(cls,
+                                    crystallized_state: CrystallizedState,
+                                    active_state: ActiveState,
+                                    block: BaseBeaconBlock,
+                                    chaindb: BaseBeaconChainDB,
+                                    cycle_length: int) -> ActiveState:
+        """
+        Process ``block`` and return the new ActiveState.
+
+
+        TODO: It doesn't match the latest spec.
+        There will be more fields need to be updated in ActiveState.
+        """
+        parent_block = chaindb.get_block_by_hash(block.parent_hash)
+        recent_block_hashes = get_new_recent_block_hashes(
+            active_state.recent_block_hashes,
+            parent_block.slot_number,
+            block.slot_number,
+            block.parent_hash
+        )
+
+        if parent_block.parent_hash != GENESIS_PARENT_HASH:
+            cls.validate_parent_block_proposer(
+                crystallized_state,
+                block,
+                parent_block,
+                cycle_length,
+            )
+        cls.validate_randao_reveal()
+
+        for attestation in block.attestations:
+            cls.validate_attestation(
+                block,
+                parent_block,
+                crystallized_state,
+                active_state,
+                attestation,
+                chaindb,
+                cycle_length,
+            )
+
+        return active_state.copy(
+            recent_block_hashes=recent_block_hashes,
+            pending_attestations=(
+                active_state.pending_attestations + block.attestations
+            ),
+        )
+
+    @classmethod
+    def compute_cycle_transitions(
+            cls,
+            crystallized_state: CrystallizedState,
+            active_state: ActiveState) -> Tuple[CrystallizedState, ActiveState]:
+        # TODO: it's a stub
+        return crystallized_state, active_state
+
+    #
+    #
+    # Proposer APIs
+    #
+    #
+    def propose_block(
+        self,
+        crystallized_state: CrystallizedState,
+        active_state: ActiveState,
+        block: BaseBeaconBlock,
+        shard_id: int,
+        shard_block_hash: Hash32,
+        chaindb: BaseBeaconBlock,
+        config: BeaconConfig,
+        private_key: int
```

Comment:
Yes, the key pairing is in ECC family. But the curve will need to be upgraded; right now we are using `BN128` in the codebase, but will have to change to [`BLS12-381`](https://z.cash/blog/new-snark-curve/).

Boneh–Lynn–Shacham (BLS) is for aggregating the signatures into one aggregated signature, and we only have to verify the aggregated signature with public keys instead of putting every single sigs on chain.